File: NET/JNetProto.Sharp/IDataReader.cs
using System;
using System.Collections;
using System.Runtime.CompilerServices;

namespace JNetProto.Sharp
{
    public interface IDataReader : IDisposable
    {
        bool ReadBool();
        byte ReadI8();
        short ReadI16();
        int ReadI32();
        long ReadI64();
        float ReadF32();
        double ReadF64();
        decimal ReadF128();
        char ReadChar();
        string ReadUtf8();
        TimeSpan ReadDuration();
        DateTime ReadTimestamp();
        Guid ReadGuid();
        Array ReadArray();
        IDictionary ReadMap();
        ITuple ReadTuple();
        IEnumerable ReadSet();
        IList ReadList();
        object[] ReadBag();
        byte[] ReadBinary();
        object ReadObject();
    }
}

File: JVM/jnetproto-java/src/main/java/jnetproto/java/IDataReader.java
package jnetproto.java;

import java.io.IOException;
import java.math.BigDecimal;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

public interface IDataReader extends AutoCloseable {
    boolean readBool() throws IOException;
    byte readI8() throws IOException;
    short readI16() throws IOException;
    int readI32() throws IOException;
    long readI64() throws IOException;
    float readF32() throws IOException;
    double readF64() throws IOException;
    BigDecimal readF128() throws IOException;
    char readChar() throws IOException;
    String readUtf8() throws IOException;
    Duration readDuration() throws IOException;
    LocalDateTime readTimestamp() throws IOException;
    UUID readGuid() throws IOException;
    Object readArray() throws IOException;
    Map readMap() throws IOException;
    Object readObject() throws IOException;
}
